### ISSUE TITLE
Importing grequests blocks socket reads forever

### ISSUE THREAD

**Issue by naktinis** [kennethreitz/grequests#51]
I'm importing grequests (which then monkey patches networking libraries) in a multi-threaded Django app run under uWSGI and I get errors like this one:

    Traceback (most recent call last):
      File "/.../python2.7/site-packages/redis/connection.py", line 182, in read_response
        buffer = self._sock.recv(4096)
      File "/.../python2.7/site-packages/gevent/socket.py", line 392, in recv
        self._wait(self._read_event)
      File "/.../python2.7/site-packages/gevent/socket.py", line 298, in _wait
        self.hub.wait(watcher)
      File "/.../python2.7/site-packages/gevent/hub.py", line 341, in wait
        result = waiter.get()
      File "/.../python2.7/site-packages/gevent/hub.py", line 568, in get
        return self.hub.switch()
      File "/.../python2.7/site-packages/gevent/hub.py", line 331, in switch
        return greenlet.switch(self)
    LoopExit: This operation would block forever

Can this be caused by reading from sockets in multiple threads spawned by uWSGI or is there another explanation for this behaviour?

**Comment by maximumG** [comment]
+1..I created a library using grequests. It took me time to understand that grequests was blocking the multiprocessing library.

**Comment by kennethreitz** [comment]
I would not recommend using grequests within a larger app, but running it as a separate process because of the global monkey-patching that gevent does.

**Comment by FaridLU** [comment]
I had faced the same issue. I couldn't find any solution with `grequests`. Then I followed this blog post: https://realpython.com/python-concurrency/ and solved the issue.

Then I made a package by following the approach of that blog, so that this helps to a lot of people, who are experiencing same issue. This package is lot more faster than `grequests` too.

Here is the package: [fgrequests][1]

I would love to hear feedback about this package. 

Thank you.


  [1]: https://pypi.org/project/fgrequests/
